File: usecases/src/main/java/com/andyanika/usecases/UseCaseModule.java
package com.andyanika.usecases;

import com.andyanika.translator.common.interfaces.usecase.AddFavoriteUseCase;
import com.andyanika.translator.common.interfaces.usecase.GetFavoritesUseCase;
import com.andyanika.translator.common.interfaces.usecase.GetLanguagesUseCase;
import com.andyanika.translator.common.interfaces.usecase.GetSelectedLanguageUseCase;
import com.andyanika.translator.common.interfaces.usecase.HistoryUseCase;
import com.andyanika.translator.common.interfaces.usecase.RemoveFavoriteUseCase;
import com.andyanika.translator.common.interfaces.usecase.SelectLanguageUseCase;
import com.andyanika.translator.common.interfaces.usecase.TranslationUseCase;

import dagger.Binds;
import dagger.Module;

@Module
public abstract class UseCaseModule {
    @Binds
    abstract TranslationUseCase provideTranslationUseCase(TranslateUseCaseImpl useCase);

    @Binds
    abstract SelectLanguageUseCase provideSelectLanguageUseCase(SelectLanguageUseCaseImpl useCase);

    @Binds
    abstract GetSelectedLanguageUseCase provideGetSelectedUseCase(GetSelectedLanguagesUseCaseImpl useCase);

    @Binds
    abstract GetLanguagesUseCase provideLanguagesUseCase(GetLanguagesUseCaseImpl useCase);

    @Binds
    abstract GetFavoritesUseCase provideGetFavoritesUseCase(GetFavoritesUseCaseImpl useCase);

    @Binds
    abstract RemoveFavoriteUseCase provideRemoveFavoritesUseCase(RemoveFavoriteUseCaseImpl useCase);

    @Binds
    abstract AddFavoriteUseCase provideAddFavoritesUseCase(AddFavoriteUseCaseImpl useCase);

    @Binds
    abstract HistoryUseCase provideHistoryUseCase(HistoryUseCaseImpl useCase);
}

File: usecases/src/main/kotlin/com/andyanika/usecases/di/koin/UseCaseModule.kt
package com.andyanika.usecases.di.koin

import com.andyanika.usecases.*
import core.interfaces.usecase.*
import org.koin.dsl.bind
import org.koin.dsl.module

internal val useCaseModule = module {
    factory { TranslateUseCaseImpl(get(), get()) } bind TranslationUseCase::class

    factory { SelectLanguageUseCaseImpl(get()) } bind SelectLanguageUseCase::class

    factory { GetSelectedLanguagesUseCaseImpl(get(), get()) } bind GetSelectedLanguageUseCase::class

    factory { GetLanguagesUseCaseImpl(get(), get()) } bind GetLanguagesUseCase::class

    factory { GetFavoritesUseCaseImpl(get()) } bind GetFavoritesUseCase::class

    factory { RemoveFavoriteUseCaseImpl(get()) } bind RemoveFavoriteUseCase::class

    factory { AddFavoriteUseCaseImpl(get()) } bind AddFavoriteUseCase::class

    factory { HistoryUseCaseImpl(get()) } bind HistoryUseCase::class
}
